sys/arm64/arm64/copyinout.S

```diff
@@ -91,28 +109,48 @@ END(copyin)
 /*
  * Copies a string from a user to kernel address
  *
- * int copyinstr(const void *udaddr, void *kaddr, size_t len, size_t *done)
+ * int copyinstr(const void * __capability udaddr, void *kaddr, size_t len,
+ *         size_t *done)
  */
 ENTRY(copyinstr)
 	mov	x5, xzr		/* count = 0 */
 	mov	w4, #1		/* If zero return faulure */
 	cbz	x2, 3f		/* If len == 0 then skip loop */
 
+#if __has_feature(capabilities)
+	cvtd	c1, x1			/* Store through a capability */
+	scbnds	c1, c1, x2		/* Set the bounds */
+#endif
+
 	adr	x6, copyio_fault /* Get the handler address */
 	SET_FAULT_HANDLER(x6, x7) /* Set the handler */
 
 	ldr	x7, =VM_MAXUSER_ADDRESS
-1:	cmp	x0, x7
+	ENTER_C64
+1:
+#if __has_feature(capabilities)
+	gcvalue	x9, c0
+	cmp	x9, x7
+#else
+	cmp	x0, x7
+#endif
 	b.cs	copyio_fault
```

Comment:
We should fix this before merging, e.g. something like (reusing ENTER_C64 as a hack to make sure we're still in c64):
```
b.cs 999f
...
ret
ENTER_C64
999:
EXIT_C64
b copyio_fault
```

Reply:
So you agree then with my comment in #740 that this is currently wrong? What's on morello-dev is probably the nicest way to do it IMO.